cloudinit/socket.py

```diff
@@ -0,0 +1,174 @@
+
+        Once the message has been received, enter the context.
+        """
+        if os.isatty(sys.stdin.fileno()):
+            LOG.info(
+                "Stdin is a tty, so skipping stage synchronization protocol"
+            )
+            return
+        self.systemd_exit_code = 0
+        sd_notify(
+            "STATUS=Waiting on external services to "
+            f"complete before starting the {self.stage} stage."
+        )
+        start_time = time.monotonic()
+        # block until init system sends us data
+        # the first value returned contains a message from the init system
+        #     (should be "start")
+        # the second value contains the path to a unix socket on which to
+        #     reply, which is expected to be /path/to/{self.stage}-return.sock
+        sock = self.sockets[self.stage]
+        chunk, self.remote = sock.recvfrom(5)
+
+        if b"start" != chunk:
+            # The protocol expects to receive a command "start"
+            self.__exit__(None, None, None)
+            raise ValueError(f"Received invalid message: [{str(chunk)}]")
+        elif f"{DEFAULT_RUN_DIR}/share/{self.stage}-return.sock" != str(
+            self.remote
+        ):
+            # assert that the return path is in a directory with appropriate
+            # permissions
+            self.__exit__(None, None, None)
+            raise ValueError(f"Unexpected path to unix socket: {self.remote}")
+
+        total = time.monotonic() - start_time
+        time_msg = f"took {total: .3f}s to " if total > 0.01 else ""
+        sd_notify(f"STATUS=Running ({self.stage} stage)")
+        LOG.debug("sync(%s): synchronization %scomplete", self.stage, time_msg)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Notify the socket that this stage is complete."""
+        message = "done"
+        if exc_type:
+            # handle exception thrown in context
+            self.systemd_exit_code = 1
+            self.experienced_any_error = True
+            status = f"{repr(exc_val)} in {exc_tb.tb_frame}"
+            message = (
+                'fatal error, run "systemctl status cloud-init-main.service" '
+                'and "cloud-init status --long" for more details'
+            )
+            if not self.first_exception:
+                self.first_exception = status
+            LOG.fatal(status)
+            sd_notify(f"STATUS={status}")
+
+        self.experienced_any_error = self.experienced_any_error or bool(
+            self.systemd_exit_code
+        )
+        sock = self.sockets[self.stage]
+        sock.connect(self.remote)
+
+        # the returned message will be executed in a subshell
+        # hardcode this message rather than sending a more informative message
+        # to avoid having to sanitize inputs (to prevent escaping the shell)
+        sock.sendall(
+            f"echo '{message}'; exit {self.systemd_exit_code};".encode()
+        )
+        sock.close()
```

Comment:
Once we finish our socket interaction, shouldn't we be removing the associate file paths to avoid leaving them in /run/cloud-init/shared/*sock files?
```suggestion
        sock.close()
        for socket_path in (self.remote, self.remote.replace("-return", "")):
            with suppress(FileNotFoundError):
                os.remove(socket_path)
```

Reply:
> Once we finish our socket interaction, shouldn't we be removing the associate file paths to avoid leaving them in /run/cloud-init/shared/*sock files?

No. Doing this would introduce a race into the socket protocol.

If we were to do that, there would be no guarantee that the receiving process was scheduled in time by the kernel to receive the message before it gets deleted.

We could "wait until the end", and then delete them all, but again we would require some way to ensure that the message was received before deletion.

These files take up very little space and are harmless, so why not just leave them?